src/shared/Microsoft.Git.CredentialManager/EnvironmentVariables.cs

```diff
@@ -0,0 +1,109 @@
+    {
+        private readonly IReadOnlyDictionary<string, string> _envars;
+
+        public EnvironmentVariables(IDictionary variables)
+        {
+            EnsureArgument.NotNull(variables, nameof(variables));
+
+            // On Windows it is technically possible to get env vars which differ only by case
+            // even though the general assumption is that they are case insensitive on Windows.
+            // For example, some of the standard .NET types like System.Diagnostics.Process
+            // will fail to start a process on Windows if given duplicate environment variables.
+            // See this issue for more information: https://github.com/dotnet/corefx/issues/13146
+
+            // If we're on the Windows platform we should de-duplicate by setting the string
+            // comparer to OrdinalIgnoreCase.
+            var comparer = PlatformUtils.IsWindows()
+                ? StringComparer.OrdinalIgnoreCase
+                : StringComparer.Ordinal;
+
+            var dict = new Dictionary<string, string>(comparer);
+
+            foreach (var key in variables.Keys)
+            {
+                if (key is string name && variables[key] is string value)
+                {
+                    dict[name] = value;
+                }
+            }
+
+            _envars = new ReadOnlyDictionary<string, string>(dict);
+        }
+
+        #region IReadOnlyDictionary<string, string>
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        public IEnumerator<KeyValuePair<string, string>> GetEnumerator() => _envars.GetEnumerator();
+        public int Count => _envars.Count;
+        public bool ContainsKey(string key) => _envars.ContainsKey(key);
+
+        /// <summary>
+        /// Try and get the value of an environment variable as a string.
+        /// </summary>
+        /// <param name="name">Environment variable name.</param>
+        /// <param name="value">Environment variable value.</param>
+        /// <returns>True if the value was found, false otherwise.</returns>
+        public bool TryGetValue(string name, out string value) => _envars.TryGetValue(name, out value);
+
+        public string this[string name] => _envars[name];
+        public IEnumerable<string> Keys => _envars.Keys;
+        public IEnumerable<string> Values => _envars.Values;
+
+        #endregion
+    }
+
+    public static class EnvironmentVariablesExtensions
+    {
+        /// <summary>
+        /// Get the value of an environment variable as a string.
+        /// </summary>
+        /// <exception cref="System.Collections.Generic.KeyNotFoundException">An environment variable with the specified name was not found.</exception>
+        /// <param name="envars">Environment variables.</param>
+        /// <param name="name">Environment variable name.</param>
+        /// <returns>Environment variable value.</returns>
+        public static string GetValue(this IEnvironmentVariables envars, string name)
```

Comment:
The docs say this throws `KeyNotFoundException` when there is no value for `name` but the implementation _looks_ like it returns `null` and doesn't throw. Which is the desired behavior?

Reply:
The exception, to be consistent with `IGitConfiguration`. Will fix.